pilot/pkg/model/context.go

```diff
@@ -160,6 +155,11 @@ func (e *Environment) Init() {
 	e.clusterLocalServices = NewClusterLocalProvider(e)
 }
 
+func (e *Environment) InitGatewayNameCache(updater XDSUpdater) (err error) {
+	e.gatewayNameCache, err = newNetworkGatewayNameCache(updater)
```

Comment:
I am a bit concerned with this synchronous flow

I think a better model may be to have something async.

* When a GW is added with hostname, we add it to 'dnsCache'. It is not effective until dnsCache has resolved.
* For all items in dnsCache, we keep them up to date in background, re-querying every once and while (influenced by TTL). If the result changes, we trigger a push.
  * One concern is the DNS response is constantly changing as a poor-mans load balancer. Hopefully no one is doing this but we may need to add some check for this...

In the current model I am a bit concerned with slow DNS updates blocking an XDS push. Also if there are no pushes for a long time, we will go long past the TTL without re-resolving

WDYT?

Reply:
Actually I see the time.AfterFunc. So my concern about not re-resolving is not valid.

IMO its better to have more DNS resolves than more pushes. The latter is a lot more expensive.

Do we know the (default) TTL of AWS LB? I am guessing its pretty low

Reply:
This sounds good, but means I'll need to refactor a few things first since we aggregate all the gateways during push-context init.

Reply:
One middle ground would be to only do synchronous lookup on the first add. This should happen rarely. Then future updates are async.

I don't think this is the best solution, but its likely way easier. And its opt in - currently anyone with Hostname is broken entirely. This doesn't negatively impact users not using hostname (right?)

Reply:
> This doesn't negatively impact users not using hostname (right?)

Right. 

>  only do synchronous lookup on the first add

We should probably refactor this to be notify-up anyway... something like: 

```go
env.ServiceDiscovery.AddNetworkGatewaysHandler(networkManager.onGatewaysChanged())
```

Reply:
> only do synchronous lookup on the first add

I had to change a ton of code to avoid this.. worth it?